### PR TITLE
Batch-verify Orchard proofs

### PR DESCRIPTION
**Pull request by str4d**
This PR also adds a new `ConnectBlock` benchmark using mainnet block 1708048, which is full of Orchard shielded transactions.

## Benchmark results

`./qa/zcash/performance-measurements.sh time connectblockorchard`

Benchmarks measured on a Ryzen 9 5950X (with the global `rayon` threadpool at its default for this CPU of 32 threads).

| Version | Avg time (s) | Decrease |
|---|---|---|
| 5.0.0 (this PR, before upgrade) | 2.41 | |
| This PR, before batching | 2.03 | 15.9% |
| This PR, after batching | 0.43 | 82.1% |

<details>
<summary>Benchmark data</summary>

### Single verification of Orchard proofs, before migrating to `orchard 0.2.0`:
```
[
  {
    "runningtime": 2.409682
  },
  {
    "runningtime": 2.422087
  },
  {
    "runningtime": 2.387994
  },
  {
    "runningtime": 2.440032
  },
  {
    "runningtime": 2.418749
  },
  {
    "runningtime": 2.395318
  },
  {
    "runningtime": 2.424455
  },
  {
    "runningtime": 2.396122
  },
  {
    "runningtime": 2.426891
  },
  {
    "runningtime": 2.396417
  }
]
```

### Single verification of Orchard proofs, after migrating to `orchard 0.2.0`:
```
[
  {
    "runningtime": 2.023175
  },
  {
    "runningtime": 2.056039
  },
  {
    "runningtime": 2.020506
  },
  {
    "runningtime": 1.981866
  },
  {
    "runningtime": 2.025942
  },
  {
    "runningtime": 2.042992
  },
  {
    "runningtime": 2.035301
  },
  {
    "runningtime": 2.023647
  },
  {
    "runningtime": 2.046091
  },
  {
    "runningtime": 2.026172
  }
]
```

### Batch verification of Orchard proofs
```
[
  {
    "runningtime": 0.4371699999999999
  },
  {
    "runningtime": 0.426709
  },
  {
    "runningtime": 0.4324750000000001
  },
  {
    "runningtime": 0.437522
  },
  {
    "runningtime": 0.418723
  },
  {
    "runningtime": 0.427963
  },
  {
    "runningtime": 0.432248
  },
  {
    "runningtime": 0.4409380000000001
  },
  {
    "runningtime": 0.423852
  },
  {
    "runningtime": 0.433306
  }
]
```
</details>

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -92,3 +92,6 @@ features = ["ansi", "env-filter", "fmt", "time"]
 lto = true
 panic = 'abort'
 codegen-units = 1
+
+[patch.crates-io]
+orchard = { git = "https://github.com/zcash/orchard.git", rev = "6c0532776c9e8ccb20af8f360c9b60a2e4d10f58" }
```
Comment (str4d) [inline Cargo.toml:97]: PR will be force-pushed to replace this once `orchard 0.2.0` is published.

Comment (str4d) [reply Cargo.toml:97]: This will also fix the failing "Cargo patches" lint.

Comment (str4d) [reply Cargo.toml:97]: And I'll fix the `cargo vet` failure at the same time.

File: qa/zcash/create_benchmark_archive.py [hunk 2]
```diff
@@ -23,9 +23,9 @@
 make -C src/leveldb/
```
Comment (daira) [inline qa/zcash/create_benchmark_archive.py:23]: This script should have a Python 3 shebang and version check.

File: src/rust/include/rust/orchard.h [hunk 3]
```diff
@@ -101,16 +86,24 @@ OrchardBatchValidatorPtr* orchard_batch_validation_init();
...
+///
+/// ## Consensus rules
+///
+/// [§4.6](https://zips.z.cash/protocol/protocol.pdf#actiondesc):
+/// - Canonical element encodings are enforced by `orchard_bundle_parse`.
+/// - SpendAuthSig^Orchard validity is enforced here.
+/// - Proof validity is enforced here.
+///
+/// [§7.1](https://zips.z.cash/protocol/protocol.pdf#txnencodingandconsensus):
+/// - `bindingSigOrchard` validity is enforced here.
```
Comment (daira) [inline src/rust/include/rust/orchard.h:106]: If I understand correctly, all of the transaction-specific consensus rules except for type and canonical encoding rules are now enforced here, so we should say that in the comment.

File: src/rust/include/rust/orchard.h [hunk 4]
```diff
@@ -160,7 +151,8 @@ class AuthValidator
         return *this;
     }
 
-    /// Creates a validation context that batch-validates Orchard signatures.
+    /// Creates a validation context that batch-validates Orchard proofs and
+    /// signatures.
```
Comment (daira) [inline src/rust/include/rust/orchard.h:155]: Isn't it validating more than that?

**Review by daira — APPROVED** [review]
utACK

**Comment by str4d** [comment]
[Force-pushed](https://github.com/zcash/zcash/compare/f9a9ee927ad2d10da2a20f8ee421198d54e919d9..b1887ac44df70c3c30b45710189ce558894a9685) to use the released `orchard 0.2.0`, and address @daira's comment.

**Review by daira — APPROVED** [review]
utACK

**Review by ebfull — APPROVED** [review]

**Review by nuttycom — APPROVED** [review]
utACK

**Comment by str4d** [comment]
[Rebased](https://github.com/zcash/zcash/compare/b1887ac44df70c3c30b45710189ce558894a9685..f4d1fbe886b0dc88d70f87c58c01fc26e3a81320) to ensure this branch included the necessary changes for CI. No changes to this PR.

**Review by daira — APPROVED** [review]
utACK (GitHub shows no changes since my last review other than the rebase).

**Review by nuttycom — APPROVED** [review]
utACK cf2a45b25a4bd004da8d4c9627a4c22add6346c0

**Review by daira — DISMISSED** [review]
utACK

**Review by daira — APPROVED** [review]
utACK (no changes since last review apart from the rebase).

**Pull request merged**